routers/repo/wiki.go

```diff
@@ -65,11 +66,12 @@ type PageMeta struct {
 // findEntryForFile finds the tree entry for a target filepath.
 func findEntryForFile(commit *git.Commit, target string) (*git.TreeEntry, error) {
 	entries, err := commit.ListEntries()
+	unescapedTarget, _ := url.QueryUnescape(target)
 	if err != nil {
 		return nil, err
 	}
 	for _, entry := range entries {
-		if entry.IsRegular() && entry.Name() == target {
+		if entry.IsRegular() && (entry.Name() == target || entry.Name() == unescapedTarget) {
```

Comment:
I'm puzzled. In Theory this should break made-up cases where both `c%C3%A1mara` and `cámara` exist, but I couldn't make it fail.

To handle this properly we should probably use *only* `unescapedTarget` (like in your previous PR) and add a migration step. However I don't think a migration is a very good idea because many things can go wrong for such an edge case.

All in all, If possible, I'd change this to check if there are any entries that matches `target`. Only if there is none then I'd resort to the one matching `unescapedTarget`.

Reply:
Now, I'm very much still learning Go, but if I understand correctly, a case where both `c%C3%A1mara` and `cámara` exists wouldn't fail because the logic stops executing after it gets a `true` for the comparison against `target`. The check against `unescapedTarget` would never even happen in that case. So it should already be performing the way you describe. I think...

Reply:
This is a for loop in a range (array), it will sequencially check values like:

```
target := "cámara"
unscapedTarget := "cámara" // "cámara" unescaped == "cámara"

aimara      ---> matches target? no; matches unscapedTarget? no
bamara      ---> matches target? no; matches unscapedTarget? no
c%C3%A1mara ---> matches target? no; matches unscapedTarget? yes --> exit from function
cámara      ---> never checked; loop stopped in previous value
```

Reply:
I'm not sure if I follow your example... Wouldn't it be more like this?

```
target := "c%C3%A1mara"
unscapedTarget := "cámara" // "c%C3%A1mara" unescaped == "cámara"

aimara        ---> matches target? no; matches unscapedTarget? no
bamara        ---> matches target? no; matches unscapedTarget? no
c%C3%A1mara   ---> matches target? yes! exit from function; matches unscapedTarget? never checked. we've already exited from the function
cámara        ---> never checked; loop stopped in previous value
```

Reply:
If the original (imported) wiki has two pages `cámara` and `c%C3%A1mara`, what are the URLs for those? According to [online encoders](https://www.urlencoder.org/) those are:
```
page1 :="c%C3%A1mara"
page2 :="c%25C3%25A1mara"
```

But then you have:

```
page1 :="c%C3%A1mara"
unescaped1 :="cámara"
page2 :="c%25C3%25A1mara"
unescaped2 :="c%C3%A1mara"
```

As you can see, `unescaped2` == `page1`. So they look the same and you have a name collision.

Reply:
(Now I know why I couldn't reproduce the problem: I had to import a wiki, not edit the wiki from Gitea's editor)

Reply:
So in that case, one of those pages would be inaccessible depending on which one comes up last in the for loop. But no matter how you write your logic, I'm pretty sure that would always be the case as long as you're effectively checking against two possible filenames. I suppose I would just assume that users aren't saving their wiki pages with machine-readable sequences of characters. Sure, doing so would cause another bug. But it seems to me like that's a lot less likely to ever be an issue than the current bug.

Reply:
>
> All in all, If possible, I'd change this to check if there are any entries that matches target. Only if there is none then I'd resort to the one matching unescapedTarget.

With a logic like this, you leave the window open for the user to fix things in case of name collition (e.g. renaming the file or changing the links).

It's actually pretty easy:

```
	entries, err := commit.ListEntries()
	if err != nil {
		return nil, err
	}
	// The longest name should be checked first
	for _, entry := range entries {
		if entry.IsRegular() && entry.Name() == target {
			return entry, nil
		}
	}
	// Then the unescaped, shortest alternative
	unescapedTarget, _ := url.QueryUnescape(target)
	for _, entry := range entries {
		if entry.IsRegular() && entry.Name() == unescapedTarget {
			return entry, nil
		}
	}
	return nil, nil
```

Reply:
Also the error from `QueryUnescape()` should not be ignored.